cpp/serve/sampler.cc

```diff
@@ -118,6 +118,95 @@ void SetProbWithArgmaxOnCPU(NDArray logits, int token_offset) {
   }
 }
 
+/*!
+ * \brief Sample a value from the input probability distribution with top-p
+ * using reject sampling. The input is a batch of distributions, and we use
+ * `unit_offset` to specify which distribution to sample from.
+ * \param prob The input batch of probability distributions.
+ * \param unit_offset The offset specifying which distribution to sample from.
+ * \param top_p The top-p value of sampling.
+ * \param uniform_sample A list of random numbers in [0, 1] for sampling, the
+ * size of the list is the maximum number of rejects we can tolerate for reject
+ * sampling.
+ * \return The sampled value.
+ * \note If we cannot find eligible choice within max_num_rejects rounds, this
+ * function will return -1.
+ */
+int RejectSampleTopPFromProb(NDArray prob, int unit_offset, double top_p,
```

Comment:
consider the following structure

```c++
// one round of reject sampling, no renormalization to avoid write
int TryRejectSampleTopPFromProb(NDArray prob, int unit_offset, float top_p, float uniform_sample) {
}
// The uniform sample NDArray is n * num_max_sample, we do this so that we can do deterministic
// parallelization and pre-populate the uniform_samples before parallel sampling
int SampleTopPFromProb(NDArray prob, int unit_offset, float top_p, NDArray uniform_samples) {
}
```